genesyscloud/resource_genesyscloud_telephony_didpool.go

```diff
@@ -0,0 +1,202 @@
+				ValidateFunc: validation.StringInSlice([]string{"PURE_CLOUD", "PURE_CLOUD_VOICE"}, false),
+			},
+		},
+	}
+}
+
+func createDidPool(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	startPhoneNumber := d.Get("start_phone_number").(string)
+	endPhoneNumber := d.Get("end_phone_number").(string)
+	description := d.Get("description").(string)
+	comments := d.Get("comments").(string)
+	poolProvider := d.Get("pool_provider").(string)
+
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	telephonyApi := platformclientv2.NewTelephonyProvidersEdgeApiWithConfig(sdkConfig)
+
+	log.Printf("Creating DID pool %s", startPhoneNumber)
+	didPool, _, err := telephonyApi.PostTelephonyProvidersEdgesDidpools(platformclientv2.Didpool{
+		StartPhoneNumber: &startPhoneNumber,
+		EndPhoneNumber:   &endPhoneNumber,
+		Description:      &description,
+		Comments:         &comments,
+		Provider:         &poolProvider,
+	})
+	if err != nil {
+		return diag.Errorf("Failed to create DID pool %s: %s", startPhoneNumber, err)
+	}
+
+	d.SetId(*didPool.Id)
+
+	log.Printf("Created DID pool %s %s", startPhoneNumber, *didPool.Id)
+	return readDidPool(ctx, d, meta)
+}
+
+func readDidPool(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	telephonyApi := platformclientv2.NewTelephonyProvidersEdgeApiWithConfig(sdkConfig)
+
+	log.Printf("Reading DID pool %s", d.Id())
+	didPool, resp, getErr := telephonyApi.GetTelephonyProvidersEdgesDidpool(d.Id())
+	if getErr != nil {
+		if resp != nil && resp.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("Failed to read DID pool %s: %s", d.Id(), getErr)
+	}
+
+	if *didPool.State == "deleted" {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("start_phone_number", *didPool.StartPhoneNumber)
+	d.Set("end_phone_number", *didPool.EndPhoneNumber)
+
+	if didPool.Description != nil {
+		d.Set("description", *didPool.Description)
+	} else {
+		d.Set("description", nil)
+	}
+
+	if didPool.Comments != nil {
+		d.Set("comments", *didPool.Comments)
+	} else {
+		d.Set("comments", nil)
+	}
+
+	if didPool.Provider != nil {
+		d.Set("pool_provider", *didPool.Provider)
+	} else {
+		d.Set("pool_provider", nil)
+	}
+
+	log.Printf("Read DID pool %s %s", d.Id(), *didPool.StartPhoneNumber)
+	return nil
+}
+
+func updateDidPool(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	startPhoneNumber := d.Get("start_phone_number").(string)
+	endPhoneNumber := d.Get("end_phone_number").(string)
+	description := d.Get("description").(string)
+	comments := d.Get("comments").(string)
+	poolProvider := d.Get("pool_provider").(string)
+
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	telephonyApi := platformclientv2.NewTelephonyProvidersEdgeApiWithConfig(sdkConfig)
+
+	log.Printf("Updating DID pool %s", d.Id())
+	_, _, err := telephonyApi.PutTelephonyProvidersEdgesDidpool(d.Id(), platformclientv2.Didpool{
```

Comment:
@ronanwatkins Why is this function returning three values?  The normal golang paradigm is to do a return value for the function call and then the err code.  Just curious.

Reply:
Hi John, I don't know about the Golang best practice but the three return values are for the returned Genesys Cloud Object, the ApiResponse, and the error.

Reply:
@PrinceMerluza That is why I directed it to Ronan.  I was curious as to why this was being done this within our SDK.  Not wrong, just curious :)